blockchain-source/src/cardano/mod.rs

```diff
@@ -286,26 +286,143 @@ async fn block_fetch(
         *last_tip_event = Instant::now();
     }
 
-    info!(%from, %tip, "making block range request");
+    if !await_on_tip && from == tip.0 {
+        info!("up to date, nothing to pull");
+        return Ok(());
+    }
 
-    let mut block_fetcher = match handle.blockfetch.request_range(from, tip.point).await? {
-        Some(block_fetcher) => block_fetcher,
-        None => {
-            debug!("no blocks found in range");
-            return Ok(());
+    let point = match handle.chainsync.request_next().await? {
+        chainsync::NextResponse::RollBackward(point, _) => point,
+        // since we did a chainsync request before, the next message should
+        // always be RollBackward to one of the points in that request (or
+        // Origin).
+        chainsync::NextResponse::RollForward(_, _) => {
+            return Err(anyhow::anyhow!("Unexpected RollForward message received")
+                .context(critical_error!()));
         }
+        chainsync::NextResponse::Await => {
+            return Err(
+                anyhow::anyhow!("Unexpected Await message received").context(critical_error!())
+            );
+        }
+    };
+
+    let (raw_header, tip) = if point == pallas_network::miniprotocols::Point::Origin
+        || point == from
+    {
+        match handle.chainsync.request_next().await? {
+            chainsync::NextResponse::RollForward(block_header, tip) => (block_header, tip),
+            chainsync::NextResponse::RollBackward(_, _) => {
+                // there shouldn't be a way to get this again after already rolling back to Origin or to the intersect point
+                return Err(anyhow::anyhow!("Unexpected RollBackward").context(critical_error!()));
+            }
+            chainsync::NextResponse::Await => match handle.chainsync.recv_while_must_reply().await?
+            {
+                chainsync::NextResponse::RollForward(block_header, _) => (block_header, tip),
+                chainsync::NextResponse::RollBackward(_, _) => return Ok(()),
+                // this is not encoded in the types, but afaik it shouldn't be possible to get this message twice in a row.
+                chainsync::NextResponse::Await => unreachable!(),
+            },
+        }
+    } else {
+        todo!();
+    };
+
+    let (slot, hash) = match raw_header.variant {
```

Comment:
@rooooooooob I was wondering if it would make sense to have this logic be in CML somehow? Or if there is already a way of writing this that I didn't think about.

Reply:
We have part of this code in the `MultiEraBlock::hash()` (in `multi_era/rust/src/utils.rs`) for  the hash part, but not the slot/difficulty part.